### PR TITLE
add upstream identifier in pkg update, get and clone(porch)

### PR DESCRIPTION
**Pull request by yuwenma**
- This PR lets `kpt` to add `internal.kpt.dev/upstream-identifier` annotation together with `kpt-merge` comment. 
- This function is called in [kpt pkg get](https://github.com/GoogleContainerTools/kpt/blob/72ef5b29cb372ed1d28e1380782651f366e2d14f/internal/util/get/get.go#L117), [kpt pkg update](https://github.com/GoogleContainerTools/kpt/blob/6008784a2a6de63ecba903d16c1a317022287b6e/internal/util/update/update.go#L185) and porch [package clone](https://github.com/GoogleContainerTools/kpt/blob/604b6a483f434d599abb2616fd5b392ee768654b/porch/pkg/engine/clone.go#L95), porch [package update](https://github.com/GoogleContainerTools/kpt/blob/604b6a483f434d599abb2616fd5b392ee768654b/porch/pkg/engine/engine.go#L532)  
- As discussed offline, we currently continue to rely on `kpt-merge` comment to match upstream resource and establish three-way merge. 

- Fix an existing issue where `addmergecomment` detects the resource meta by kyaml `IsNilOrEmpty`. not the meta can never be NilorEmpty since helper annotations are always added.

### CODE REVIEW

**Comment by yuwenma** [comment]
Can you review this change @justinsb @droot

File: internal/util/addmergecomment/addmergecomment.go [hunk 1]
```diff
@@ -49,6 +62,35 @@ func Process(paths ...string) error {
 	return nil
 }
 
+// addUpstreamAnnotation adds internal.kpt.dev/upstream-identifier annotation to resource.
+func addUpstreamAnnotation(object *kyaml.RNode, mergeComment string) error {
+	annotations := object.GetAnnotations()
+	if _, ok := annotations[upstreamIdentifier]; ok {
+		// skip adding annotation if the upstream annotation is already present
```
Comment (droot) [inline internal/util/addmergecomment/addmergecomment.go:69]: If I clone a package that already has these annotations (that is a case for any 3 level pkg chain), this will skip adding annotation ? with this, all descendant packages will have GVKNN of the resources in the root pkg.

File: internal/util/addmergecomment/addmergecomment.go [hunk 2]
```diff
@@ -60,15 +102,33 @@ func (amc *AddMergeComment) Filter(object *kyaml.RNode) (*kyaml.RNode, error) {
...
+		//    internal.config.kubernetes.io/path: 'k8s-cli-982798852.yaml'
+		//    internal.config.kubernetes.io/seqindent: 'compact'
+		//    internal.config.kubernetes.io/annotations-migration-resource-id: '0'
 		return object, nil
 	}
-	if strings.Contains(mf.Key.YNode().LineComment, merge.MergeCommentPrefix) {
-		// skip adding merge comment if merge comment is already present
+
+	// Only add merge comment if merge comment does not present
+	if !strings.Contains(mf.Key.YNode().LineComment, merge.MergeCommentPrefix) {
```
Comment (droot) [inline internal/util/addmergecomment/addmergecomment.go:124]: hmm... I thought the merge-key comment tracks the immediate upstream resource's identifier, but looking this it tracks the root resource's identifier.

Two options:
1) preserve existing behavior (merge-key tracks root resources'id)
2) or change it to use immediate upstream

Both of these will work for update, only difference is the number of descendant nodes affected when root change's name/namespace.

Comment (yuwenma) [reply internal/util/addmergecomment/addmergecomment.go:124]: I think both options can give identifier functions reliable id and both are more reasonable than the other in certain scenarios. So it defers to future plans on which scenarios are more important.
I'll switch to immediate upstream and list the pros and cons on different options in the PR description.

File: internal/util/addmergecomment/addmergecomment.go (hunk 1 above)
Comment (yuwenma) [reply internal/util/addmergecomment/addmergecomment.go]: I changed the code to use immediate upstream meta as the origin. Also added a comment to mark the potential caveats in 3-way-merge

File: porch/pkg/engine/engine.go [hunk 3]
```diff
@@ -21,6 +21,7 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"reflect"
```
Comment (mengqiy) [inline porch/pkg/engine/engine.go:24]: This change seems to be from another commit in a different PR. I suspect something messed up with rebase.

**Review by mengqiy — COMMENTED** [review]
LGTM if others' comments are addressed.
I will deter to Sunil or Justin for the final approval

**Review by droot — APPROVED** [review]
LGTM.

Would be good to add any tests for a 3-level pkg chain if we can.

File: porch/pkg/engine/engine.go (hunk 3 above)
Comment (yuwenma) [reply porch/pkg/engine/engine.go]: Good catch! Resolved by another rebase.
